pkg/search/postgres/joins_test.go

```diff
@@ -0,0 +1,613 @@
+					}: {
+						{
+							ColumnNameInThisSchema:  "t1_c1",
+							ColumnNameInOtherSchema: "t3_c1",
+						},
+						{
+							ColumnNameInThisSchema:  "t1_c2",
+							ColumnNameInOtherSchema: "t3_c2",
+						},
+					},
+				},
+			},
+		},
+		"required filed will not remove table": {
+			ReachableFields: map[string]searchFieldMetadata{
+				"test": {
+					baseField: &walker.Field{
+						Schema: &walker.Schema{
+							Table: "t2",
+						},
+					},
+					derivedMetadata: nil,
+				},
+			},
+			Root: &joinTreeNode{
+				currNode: &walker.Schema{
+					Table: "t1",
+				},
+				children: map[*joinTreeNode][]walker.ColumnNamePair{
+					{
+						currNode: &walker.Schema{
+							Table: "t2",
+						},
+						children: map[*joinTreeNode][]walker.ColumnNamePair{
+							{
+								currNode: &walker.Schema{
+									Table: "t3",
+								},
+								children: make(map[*joinTreeNode][]walker.ColumnNamePair),
+							}: {
+								{
+									ColumnNameInThisSchema:  "t2_c1",
+									ColumnNameInOtherSchema: "t3_c1",
+								},
+							},
+						},
+					}: {
+						{
+							ColumnNameInThisSchema:  "t1_c1",
+							ColumnNameInOtherSchema: "t2_c1",
+						},
+					},
+				},
+			},
+			ExpectedRoot: &joinTreeNode{
+				currNode: &walker.Schema{
+					Table: "t1",
+				},
+				children: map[*joinTreeNode][]walker.ColumnNamePair{
+					{
+						currNode: &walker.Schema{
+							Table: "t2",
+						},
+						children: map[*joinTreeNode][]walker.ColumnNamePair{
+							{
+								currNode: &walker.Schema{
+									Table: "t3",
+								},
+								children: make(map[*joinTreeNode][]walker.ColumnNamePair),
+							}: {
+								{
+									ColumnNameInThisSchema:  "t2_c1",
+									ColumnNameInOtherSchema: "t3_c1",
+								},
+							},
+						},
+					}: {
+						{
+							ColumnNameInThisSchema:  "t1_c1",
+							ColumnNameInOtherSchema: "t2_c1",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	return data
+}
+
+func TestRemoveUnnecessaryRelations(t *testing.T) {
+	t.Parallel()
+
+	testData := getTestData()
+	for testName, innerTestRecord := range testData {
+		t.Run(testName, func(t *testing.T) {
+			innerTestRecord.Root.removeUnnecessaryRelations(innerTestRecord.ReachableFields)
+
+			expectedInnerJoins := innerTestRecord.ExpectedRoot.toInnerJoins()
+			innerJoins := innerTestRecord.Root.toInnerJoins()
+
+			// We have to sort before using DeepEqual. Otherwise, results could defer.
+			// It's sufficient to sort by joined tables.
+			sort.SliceStable(expectedInnerJoins, func(i, j int) bool {
+				return expectedInnerJoins[i].rightTable+expectedInnerJoins[i].leftTable < expectedInnerJoins[j].rightTable+expectedInnerJoins[j].leftTable
+			})
+			sort.SliceStable(innerJoins, func(i, j int) bool {
+				return innerJoins[i].rightTable+innerJoins[i].leftTable < innerJoins[j].rightTable+innerJoins[j].leftTable
+			})
+
+			assert.True(t, reflect.DeepEqual(expectedInnerJoins, innerJoins))
```

Comment:
```suggestion
			assert.EqualValues(t, expectedInnerJoins, innerJoins)
```

Reply:
We still have to sort. The reason for that is that "children" inside joinTreeNode is a map where a pointer is used as a key. Because of that order in a loop (range) can be different.

I have updated the comment and changed it to `assert.EqualValues` - it gives better output.